File: searchlib/src/main/java/com/yahoo/searchlib/aggregation/ArgminAggregationResult.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.searchlib.aggregation;

import com.yahoo.searchlib.expression.ExpressionNode;
import com.yahoo.searchlib.expression.FloatResultNode;
import com.yahoo.searchlib.expression.ResultNode;
import com.yahoo.searchlib.expression.SingleResultNode;
import com.yahoo.vespa.objects.Deserializer;
import com.yahoo.vespa.objects.ObjectVisitor;
import com.yahoo.vespa.objects.Serializer;

/**
 * This is an aggregated result holding the value of the aggregating expression for the first hit in the group.
 *
 * Which hit is first is decided by an ordering key: with a key expression set, the first hit is the one with the
 * smallest key (negate the key to select by descending order); without one, the key is the negated hit rank, so the
 * first hit is the highest ranked one. The key of the winning hit is part of the result, so that merging partial
 * results from several nodes picks the same hit no matter what order they merge in. Ties keep the candidate seen
 * first, which is unspecified across nodes.
 *
 * @author johsol
 */
public class FirstAggregationResult extends AggregationResult {

    public static final int classId = registerClass(0x4000 + 183, FirstAggregationResult.class, FirstAggregationResult::new);
    private ExpressionNode keyExpression = null;
    private SingleResultNode key = new FloatResultNode(0.0);
    private SingleResultNode first = new FloatResultNode(0.0);
    private boolean hasValue = false;

    /**
     * Constructs an empty result node.
     */
    public FirstAggregationResult() {}

    /**
     * Constructs an instance of this class with the given value, keyed on nothing.
     *
     * @param first The value of the first hit.
     */
    public FirstAggregationResult(SingleResultNode first) {
        setFirst(first);
    }

    /**
     * Constructs an instance of this class with the given value and ordering key.
     *
     * @param first The value of the first hit.
     * @param key The ordering key of the first hit.
     */
    public FirstAggregationResult(SingleResultNode first, SingleResultNode key) {
        setFirst(first);
        setKey(key);
    }

    /**
     * Returns the value of the first hit in this.
     *
     * @return The value of the first hit.
     */
    public final SingleResultNode getFirst() {
        return first;
    }

    /**
     * Sets the value of the first hit in this, marking this as holding a value.
     *
     * @param first The value to set.
     * @return This, to allow chaining.
     */
    public final FirstAggregationResult setFirst(SingleResultNode first) {
        this.first = first;
        this.hasValue = true;
        return this;
    }

    /**
     * Returns the ordering key of the hit the value was taken from.
     *
     * @return The ordering key.
     */
    public final SingleResultNode getKey() {
        return key;
    }

    /**
     * Sets the ordering key of the hit the value was taken from.
     *
     * @param key The ordering key to set.
     * @return This, to allow chaining.
     */
    public final FirstAggregationResult setKey(SingleResultNode key) {
        this.key = key;
        return this;
    }

    /**
     * Returns the expression producing the ordering key, or null when ordering by hit rank.
     *
     * @return The key expression, or null.
     */
    public final ExpressionNode getKeyExpression() {
        return keyExpression;
    }

    /**
     * Sets the expression producing the ordering key.
     *
     * @param keyExpression The key expression to set.
     * @return This, to allow chaining.
     */
    public final FirstAggregationResult setKeyExpression(ExpressionNode keyExpression) {
        this.keyExpression = keyExpression;
        return this;
    }

    /**
     * Returns whether any hit has been aggregated into this.
     *
     * @return True if this holds the value of a hit.
     */
    public final boolean hasValue() {
        return hasValue;
    }

    @Override
    public ResultNode getRank() {
        return first;
    }

    @Override
    protected int onGetClassId() {
        return classId;
    }

    @Override
    protected void onSerialize(Serializer buf) {
        super.onSerialize(buf);
        serializeOptional(buf, keyExpression);
        buf.putByte(null, hasValue ? (byte)1 : (byte)0);
        serializeOptional(buf, key);
        serializeOptional(buf, first);
    }

    @Override
    protected void onDeserialize(Deserializer buf) {
        super.onDeserialize(buf);
        keyExpression = (ExpressionNode)deserializeOptional(buf);
        hasValue = buf.getByte(null) != 0;
        key = (SingleResultNode)deserializeOptional(buf);
        first = (SingleResultNode)deserializeOptional(buf);
    }

    @Override
    protected void onMerge(AggregationResult result) {
        FirstAggregationResult rhs = (FirstAggregationResult)result;
        if (!rhs.hasValue || (hasValue && rhs.key.compareTo(key) >= 0)) {
            return;
        }
        first = (SingleResultNode)rhs.first.clone();
        key = (SingleResultNode)rhs.key.clone();
        hasValue = true;
    }

    @Override
    public FirstAggregationResult clone() {
        FirstAggregationResult obj = (FirstAggregationResult)super.clone();
        if (keyExpression != null) {
            obj.keyExpression = keyExpression.clone();
        }
        if (key != null) {
            obj.key = (SingleResultNode)key.clone();
        }
        if (first != null) {
            obj.first = (SingleResultNode)first.clone();
        }
        return obj;
    }

    @Override
    protected boolean equalsAggregation(AggregationResult obj) {
        FirstAggregationResult rhs = (FirstAggregationResult)obj;
        return hasValue == rhs.hasValue && equals(keyExpression, rhs.keyExpression) && equals(key, rhs.key)
                && equals(first, rhs.first);
    }

    @Override
    public void visitMembers(ObjectVisitor visitor) {
        super.visitMembers(visitor);
        visitor.visit("keyExpression", keyExpression);
        visitor.visit("hasValue", hasValue);
        visitor.visit("key", key);
        visitor.visit("first", first);
    }
}

File: searchlib/src/vespa/searchlib/aggregation/argmin_aggregation_result.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "argmin_aggregation_result.h"

#include <vespa/searchlib/expression/floatresultnode.h>
#include <vespa/searchlib/expression/resultvector.h>

#include <vespa/searchlib/aggregation/aggregation.hpp>
#include <vespa/vespalib/objects/visit.hpp>

namespace search::aggregation {

using expression::ExpressionTree;
using expression::FloatResultNode;
using expression::ResultNode;
using expression::ResultNodeVector;
using expression::SingleResultNode;

using vespalib::Deserializer;
using vespalib::Serializer;

IMPLEMENT_AGGREGATIONRESULT(ArgminAggregationResult, AggregationResult);

ArgminAggregationResult::ArgminAggregationResult()
    : AggregationResult(),
      _key_tree(std::make_shared<ExpressionTree>()),
      _key(FloatResultNode(0.0)),
      _scratch_key(FloatResultNode(0.0)),
      _value(FloatResultNode(0.0)),
      _has_value(false) {
}

ArgminAggregationResult::ArgminAggregationResult(const SingleResultNode& key, const SingleResultNode& value)
    : AggregationResult(),
      _key_tree(std::make_shared<ExpressionTree>()),
      _key(key),
      _scratch_key(key),
      _value(value),
      _has_value(true) {
}

ArgminAggregationResult::~ArgminAggregationResult() = default;

ArgminAggregationResult& ArgminAggregationResult::set_key_expression(ExpressionNode::UP key) {
    _key_tree = std::make_shared<ExpressionTree>(std::move(key));
    // Pick up the key type right away when the value expression is already prepared.
    const ExpressionNode* value_expr = getExpression();
    if (value_expr != nullptr && value_expr->getResult() != nullptr) {
        onPrepare(*value_expr->getResult());
    }
    return *this;
}

void ArgminAggregationResult::onPrepare(const ResultNode& result) {
    if (!is_ready(_value.get(), result)) {
        _value = create_and_ensure_wanted<SingleResultNode, FloatResultNode>(result);
    }
    const ResultNode* key_result = (_key_tree->getRoot() != nullptr) ? _key_tree->getResult() : nullptr;
    if (key_result != nullptr && !is_ready(_key.get(), *key_result)) {
        _key = create_and_ensure_wanted<SingleResultNode, FloatResultNode>(*key_result);
    } else if (_key.get() == nullptr) {
        _key = std::make_unique<FloatResultNode>();
    }
    if (!is_ready(_scratch_key.get(), *_key)) {
        _scratch_key.reset(static_cast<SingleResultNode*>(_key->getClass().create()));
    }
}

void ArgminAggregationResult::initForUnitTest(const ResultNode& result) {
    onPrepare(result);
    _value->set(result);
    _has_value = true;
}

bool ArgminAggregationResult::fill_scratch_key() {
    const ResultNode* key = (_key_tree->getRoot() != nullptr) ? _key_tree->getResult() : nullptr;
    if (key == nullptr) {
        return false;
    }
    if (key->isMultiValue()) {
        const auto& values = static_cast<const ResultNodeVector&>(*key);
        if (values.empty()) {
            return false;
        }
        _scratch_key->set(values.get(0));
    } else {
        _scratch_key->set(*key);
    }
    return true;
}

void ArgminAggregationResult::update_value(const ResultNode& result) {
    if (_has_value && _scratch_key->cmp(*_key) >= 0) {
        return;
    }
    if (result.isMultiValue()) {
        const auto& values = static_cast<const ResultNodeVector&>(result);
        if (values.empty()) {
            return;
        }
        _value->set(values.get(0));
    } else {
        _value->set(result);
    }
    _key->set(*_scratch_key);
    _has_value = true;
}

void ArgminAggregationResult::onAggregate(const ResultNode& result) {
    if (fill_scratch_key()) {
        update_value(result);
    }
}

void ArgminAggregationResult::onAggregate(const ResultNode& result, DocId docId, HitRank rank) {
    if (_key_tree->getRoot() != nullptr) {
        _key_tree->execute(docId, rank);
    }
    onAggregate(result);
}

void ArgminAggregationResult::onAggregate(const ResultNode& result, const document::Document& doc, HitRank rank) {
    if (_key_tree->getRoot() != nullptr) {
        _key_tree->execute(doc, rank);
    }
    onAggregate(result);
}

void ArgminAggregationResult::onMerge(const AggregationResult& b) {
    const auto& rhs = static_cast<const ArgminAggregationResult&>(b);
    if (!rhs._has_value || (_has_value && rhs._key->cmp(*_key) >= 0)) {
        return;
    }
    _value = rhs._value;
    _key = rhs._key;
    _has_value = true;
}

void ArgminAggregationResult::onReset() {
    _value.reset(static_cast<SingleResultNode*>(_value->getClass().create()));
    _key.reset(static_cast<SingleResultNode*>(_key->getClass().create()));
    _has_value = false;
}

Serializer& ArgminAggregationResult::onSerialize(Serializer& os) const {
    AggregationResult::onSerialize(os);
    os << *_key_tree;
    return os << _has_value << _key << _value;
}

Deserializer& ArgminAggregationResult::onDeserialize(Deserializer& is) {
    AggregationResult::onDeserialize(is);
    _key_tree = std::make_shared<ExpressionTree>();
    is >> *_key_tree;
    return is >> _has_value >> _key >> _value;
}

void ArgminAggregationResult::visitMembers(vespalib::ObjectVisitor& visitor) const {
    AggregationResult::visitMembers(visitor);
    visit(visitor, "keyExpression", _key_tree);
    visit(visitor, "hasValue", _has_value);
    visit(visitor, "key", _key);
    visit(visitor, "value", _value);
}

void ArgminAggregationResult::selectMembers(const vespalib::ObjectPredicate& predicate,
                                            vespalib::ObjectOperation&       operation) {
    AggregationResult::selectMembers(predicate, operation);
    _key_tree->select(predicate, operation);
}

} // namespace search::aggregation
